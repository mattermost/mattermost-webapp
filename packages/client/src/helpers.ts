// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export function buildQueryString(parameters: Record<string, any>): string {
    const keys = Object.keys(parameters);
    if (keys.length === 0) {
        return '';
    }

    const queryParams = Object.entries(parameters).
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        filter(([_, value]) => value !== undefined).
        map(([key, value]) => `${key}=${encodeURIComponent(value)}`).
        join('&');

    return queryParams.length > 0 ? `?${queryParams}` : '';
}
