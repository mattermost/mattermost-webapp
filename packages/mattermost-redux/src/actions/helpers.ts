// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Client4} from 'mattermost-redux/client';
import {UserTypes} from 'mattermost-redux/action_types';

import {Client4Error} from 'mattermost-redux/types/client4';
import {batchActions, Action, ActionFunc, GenericAction, DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';

import {logError} from './errors';
type ActionType = string;
const HTTP_UNAUTHORIZED = 401;
export function forceLogoutIfNecessary(err: Client4Error, dispatch: DispatchFunc, getState: GetStateFunc) {
    const {currentUserId} = getState().entities.users;

    if ('status_code' in err && err.status_code === HTTP_UNAUTHORIZED && err.url && err.url.indexOf('/login') === -1 && currentUserId) {
        Client4.setToken('');
        dispatch({type: UserTypes.LOGOUT_SUCCESS, data: {}});
    }
}

function dispatcher(type: ActionType, data: any, dispatch: DispatchFunc) {
    if (type.indexOf('SUCCESS') === -1) { // we don't want to pass the data for the request types
        dispatch(requestSuccess(type, data));
    } else {
        dispatch(requestData(type));
    }
}

export function requestData(type: ActionType): GenericAction {
    return {
        type,
        data: null,
    };
}

export function requestSuccess(type: ActionType, data: any) {
    return {
        type,
        data,
    };
}

export function requestFailure(type: ActionType, error: Client4Error): any {
    return {
        type,
        error,
    };
}

/**
 * Returns an ActionFunc which calls a specfied (client) function and
 * dispatches the specifed actions on request, success or failure.
 *
 * @export
 * @param {Object} obj                                       an object for destructirung required properties
 * @param {() => Promise<mixed>} obj.clientFunc              clientFunc to execute
 * @param {ActionType} obj.onRequest                         ActionType to dispatch on request
 * @param {(ActionType | Array<ActionType>)} obj.onSuccess   ActionType to dispatch on success
 * @param {ActionType} obj.onFailure                         ActionType to dispatch on failure
 * @param {...Array<any>} obj.params
 * @returns {ActionFunc} ActionFunc
 */

export function bindClientFunc({
    clientFunc,
    onRequest,
    onSuccess,
    onFailure,
    params = [],
}: {
    clientFunc: (...args: any[]) => Promise<any>;
    onRequest?: ActionType;
    onSuccess?: ActionType | ActionType[];
    onFailure?: ActionType;
    params?: any[];
}): ActionFunc {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        if (onRequest) {
            dispatch(requestData(onRequest));
        }

        let data: any = null;
        try {
            data = await clientFunc(...params);
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            const actions: Action[] = [logError(error)];
            if (onFailure) {
                actions.push(requestFailure(onFailure, error));
            }
            dispatch(batchActions(actions));
            return {error};
        }

        if (Array.isArray(onSuccess)) {
            onSuccess.forEach((s) => {
                dispatcher(s, data, dispatch);
            });
        } else if (onSuccess) {
            dispatcher(onSuccess, data, dispatch);
        }

        return {data};
    };
}

// Debounce function based on underscores modified to use es6 and a cb

export function debounce(func: (...args: any) => unknown, wait: number, immediate: boolean, cb: () => unknown) {
    let timeout: NodeJS.Timeout|null;
    return function fx(...args: any[]) {
        const runLater = () => {
            timeout = null;
            if (!immediate) {
                Reflect.apply(func, null, args);
                if (cb) {
                    cb();
                }
            }
        };
        const callNow = immediate && !timeout;
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(runLater, wait);
        if (callNow) {
            Reflect.apply(func, null, args);
            if (cb) {
                cb();
            }
        }
    };
}

export class FormattedError extends Error {
    intl: {
        id: string;
        defaultMessage: string;
        values: any;
    };

    constructor(id: string, defaultMessage: string, values: any = {}) {
        super(defaultMessage);
        this.intl = {
            id,
            defaultMessage,
            values,
        };
    }
}

