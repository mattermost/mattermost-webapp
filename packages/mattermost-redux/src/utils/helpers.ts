// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import shallowEqual from 'shallow-equals';

import * as reselect from 'reselect';

// eslint-disable-next-line @typescript-eslint/ban-types
export function memoizeResult<F extends Function>(func: F, measure: Function | undefined = undefined): F {
    let lastArgs: IArguments|null = null;
    let lastResult: any = null;

    // we reference arguments instead of spreading them for performance reasons
    return function memoizedFunc() {
        if (!shallowEqual(lastArgs, arguments)) { //eslint-disable-line prefer-rest-params
            //eslint-disable-line prefer-rest-params
            // apply arguments instead of spreading for performance.
            const result = Reflect.apply(func, null, arguments); //eslint-disable-line prefer-rest-params
            if (!shallowEqual(lastResult, result)) {
                lastResult = result;
            }
        }

        if (measure) {
            measure();
        }

        lastArgs = arguments; //eslint-disable-line prefer-rest-params
        return lastResult;
    } as unknown as F;
}

// Use this selector when you want a shallow comparison of the arguments and you want to memoize the result
// try and use this only when your selector returns an array of ids
export const createIdsSelector = reselect.createSelectorCreator(memoizeResult);

// Use this selector when you want a shallow comparison of the arguments and you don't need to memoize the result
export const createShallowSelector = reselect.createSelectorCreator(reselect.defaultMemoize, shallowEqual as any);

// Generates a RFC-4122 version 4 compliant globally unique identifier.
export function generateId(): string {
    // implementation taken from http://stackoverflow.com/a/2117523
    let id = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
    id = id.replace(/[xy]/g, (c) => {
        const r = Math.floor(Math.random() * 16);
        let v;

        if (c === 'x') {
            v = r;
        } else {
            // eslint-disable-next-line no-mixed-operators
            v = r & 0x3 | 0x8;
        }

        return v.toString(16);
    });
    return id;
}

export function isEmail(email: string): boolean {
    // writing a regex to match all valid email addresses is really, really hard. (see http://stackoverflow.com/a/201378)
    // this regex ensures:
    // - at least one character that is not a space, comma, or @ symbol
    // - followed by a single @ symbol
    // - followed by at least one character that is not a space, comma, or @ symbol
    // this prevents <Outlook Style> outlook.style@domain.com addresses and multiple comma-separated addresses from being accepted
    return (/^[^ ,@]+@[^ ,@]+$/).test(email);
}
