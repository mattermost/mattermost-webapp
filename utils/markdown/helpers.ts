// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

// parseImageDimensions takes an href returned by marked and separates out any image dimensions if those are specified.
// If the input contains dimensions, the returned href will have them removed and they will be returned as strings,
// otherwise, the returned dimensions will both be the empty string. If only one dimension is provided, the other will
// be the string "auto".
export function parseImageDimensions(input: string): {href: string; height: string; width: string} {
    const match = (/ =(\d*)(?:x(\d+))?$/).exec(input);
    if (!match) {
        return {
            href: input,
            height: '',
            width: '',
        };
    }

    let width = match[1];
    let height = match[2];
    if (!width && !height) {
        return {
            href: input,
            height: '',
            width: '',
        };
    }

    if (width && !height) {
        height = 'auto';
    } else if (height && !width) {
        width = 'auto';
    }

    return {
        href: input.substring(0, input.length - match[0].length),
        height,
        width,
    };
}
