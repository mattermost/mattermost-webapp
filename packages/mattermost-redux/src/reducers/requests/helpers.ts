// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {RequestStatus} from 'mattermost-redux/constants';
import {GenericAction} from 'mattermost-redux/types/actions';
import {RequestStatusType} from 'mattermost-redux/types/requests';
export function initialRequestState(): RequestStatusType {
    return {
        status: RequestStatus.NOT_STARTED,
        error: null,
    };
}

export function handleRequest(
    REQUEST: string,
    SUCCESS: string,
    FAILURE: string,
    state: RequestStatusType,
    action: GenericAction,
): RequestStatusType {
    switch (action.type) {
    case REQUEST:
        return {
            ...state,
            status: RequestStatus.STARTED,
        };
    case SUCCESS:
        return {
            ...state,
            status: RequestStatus.SUCCESS,
            error: null,
        };
    case FAILURE: {
        return {
            ...state,
            status: RequestStatus.FAILURE,
            error: action.error,
        };
    }
    default:
        return state;
    }
}
