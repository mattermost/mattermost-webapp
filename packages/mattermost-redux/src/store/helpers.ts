// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {combineReducers, AnyAction} from 'redux';

import {General} from '../constants';

import {enableBatching, Action, Reducer} from 'mattermost-redux/types/actions';
import {GlobalState} from 'mattermost-redux/types/store';
import deepFreezeAndThrowOnMutation from 'mattermost-redux/utils/deep_freeze';

import reducerRegistry from './reducer_registry';

export function createReducer(baseState: GlobalState, ...reducers: Reducer[]): Reducer<GlobalState, Action> {
    reducerRegistry.setReducers(Object.assign({}, ...reducers));
    const baseReducer = combineReducers(reducerRegistry.getReducers());

    // Root reducer wrapper that listens for reset events.
    // Returns whatever is passed for the data property
    // as the new state.
    const offlineReducer = (state: GlobalState, action: Action) => {
        if ('type' in action && 'data' in action && action.type === General.OFFLINE_STORE_RESET) {
            return baseReducer(action.data || baseState, action);
        }

        return baseReducer(state, action as AnyAction);
    };

    return enableFreezing(enableBatching(offlineReducer as Reducer<GlobalState, Action>));
}

function enableFreezing(reducer: Reducer) {
    // Skip the overhead of freezing in production.
    // eslint-disable-next-line no-process-env
    if (process.env.NODE_ENV === 'production') {
        return reducer;
    }

    const frozenReducer = (state: GlobalState, action: Action) => {
        const nextState = reducer(state, action);

        if (nextState !== state) {
            deepFreezeAndThrowOnMutation(nextState);
        }

        return nextState;
    };

    return frozenReducer as Reducer<GlobalState, Action>;
}
