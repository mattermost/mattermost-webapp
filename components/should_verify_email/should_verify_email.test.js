// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {shallowWithIntl} from 'tests/helpers/intl-test-helper';

import ShouldVerifyEmail from './should_verify_email';

describe('components/ShouldVerifyEmail', () => {
    const baseProps = {
        location: {
            search: '',
        },
        siteName: 'Mattermost',
        actions: {
            sendVerificationEmail: jest.fn().mockResolvedValue({data: true}),
        },
    };

    it('should match snapshot', () => {
        const wrapper = shallowWithIntl(<ShouldVerifyEmail {...baseProps}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should call the sendVerificationEmail() action on click the button', () => {
        const props = {
            ...baseProps,
            location: {
                search: '?email=test@example.com',
            },
        };

        const wrapper = shallowWithIntl(<ShouldVerifyEmail {...props}/>);
        wrapper.find('button').simulate('click');

        expect(props.actions.sendVerificationEmail).toHaveBeenCalledWith('test@example.com');
    });
});
