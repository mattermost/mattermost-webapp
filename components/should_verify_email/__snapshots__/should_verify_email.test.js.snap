// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ShouldVerifyEmail should match snapshot 1`] = `
<div>
  <BackButton
    url="/"
  />
  <div
    className="col-sm-12"
  >
    <div
      className="signup-team__container"
    >
      <h3>
        <FormattedMessage
          defaultMessage="{siteName}: You are almost done"
          id="email_verify.almost"
          values={
            Object {
              "siteName": "Mattermost",
            }
          }
        />
      </h3>
      <div>
        <p>
          <FormattedMessage
            defaultMessage="Please verify your email address. Check your inbox for an email."
            id="email_verify.notVerifiedBody"
            values={Object {}}
          />
        </p>
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          <FormattedMessage
            defaultMessage="Resend Email"
            id="email_verify.resend"
            values={Object {}}
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
